helper/validation/validation.go

```diff
@@ -105,6 +106,16 @@ func StringLenBetween(min, max int) schema.SchemaValidateFunc {
 	}
 }
 
+// NoZeroValues is a SchemaValidateFunc which tests if the provided value is
+// not a zero value. It's useful in situations where you want to catch
+// explicit zero values on things like required fields during validation.
+func NoZeroValues(i interface{}, k string) (s []string, es []error) {
+	if reflect.ValueOf(i).Interface() == reflect.Zero(reflect.TypeOf(i)).Interface() {
+		es = append(es, fmt.Errorf("%s cannot be empty or zero", k))
```

Comment:
If I'm reading this correctly, it looks like what is disallowed depends on the value type, which in turn depends on the schema `Type`.

With that in mind, it's a little weird to respond to an empty string by saying "cannot be empty or zero", since in that case it _can_ be zero in the sense that `"0"` is not the zero value for a string.

Could we switch here based on `reflect.TypeOf(i).Kind` so we can produce a more specific error message?

```go
switch reflect.TypeOf(i).Kind {
case reflect.String:
    es = append(es, fmt.Errorf("%s must not be empty", k))
case reflect.Int, reflect.Float64:
    es = append(es, fmt.Errorf("%s must not be zero", k))
default:
    // this validator should only ever be applied to TypeString, TypeInt and TypeFloat
    panic(fmt.Errorf("can't use NoZeroValues with %T attribute %s", i, k))
}
```

Also, stylistically I think it's better to say "must not" rather than "cannot", since we're talking about a constraint being imposed rather than an innate characteristic of the type. (It _can_ be empty, but it _shouldn't_ be.)